### PR TITLE
policy-engine/src/main: Set basic CORS headers

### PR DESCRIPTION
**Pull request by wking**
Allow [simple requests][1] by including the:

```
Access-Control-Allow-Origin: *
```

header, so folks can write cross-origin web applications that query our Cincinnati server.  We're going to be building the response anyway, so we don't even get denial-of-service protection by blocking cross-origin requests for Cincy data.

Also set [`Access-Control-Allow-Methods`][2] for good measure, although we do not need to support preflight requests.

Implementation is based on [these docs][3].

[1]: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#Simple_requests
[2]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Methods
[3]: https://docs.rs/actix-web/0.5.8/actix_web/middleware/cors/

### CODE REVIEW

**Comment by wking** [comment]
[cargo-test][1]:

```
 41 | use actix_web::middleware::cors::Cors;
   |                            ^^^^ could not find `cors` in `middleware`
```

because CORS support moved out of actix-web into its own crate [in actix-web 1.0.2][2].  Pushed v906508b -> 8c8067f to pivot to actix-cors ^0.2.0.

[1]: https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_cincinnati/196/pull-ci-openshift-cincinnati-master-cargo-test/702#1:build-log.txt%3A675
[2]: https://docs.rs/crate/actix-web/1.0.2/source/CHANGES.md

**Comment by wking** [comment]
[cargo-test][1]:

```
 error[E0277]: the trait bound `actix_cors::CorsFactory: actix_service::transform::IntoTransform<_, actix_web::app_service::AppRouting>` is not satisfied
  --> policy-engine/src/main.rs:85:17
   |
85 | /                 Cors::new()
86 | |                     .allowed_origin("*")
87 | |                     .allowed_methods(vec!["HEAD", "GET"])
88 | |                     .finish(),
   | |_____________________________^ the trait `actix_service::transform::IntoTransform<_, actix_web::app_service::AppRouting>` is not implemented for `actix_cors::CorsFactory`
```

That also turns up [here][2], but I'm not clear enough on the Rust to understand what they're saying there yet.  My example looks a lot like [the actix-cors docs][3] to me.

[1]: https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_cincinnati/196/pull-ci-openshift-cincinnati-master-cargo-test/707
[2]: https://github.com/actix/actix-web/issues/934#issuecomment-546983746
[3]: https://docs.rs/actix-cors/0.2.0/actix_cors/

File: policy-engine/Cargo.toml
Comment (steveej) [inline policy-engine/Cargo.toml]:
```suggestion
actix-cors = "^0.1.0"
```

File: policy-engine/src/main.rs
Comment (steveej) [inline policy-engine/src/main.rs]:
```suggestion
```

**Review by steveej — CHANGES_REQUESTED** [review]
The used versions of `actix-cors` and `actix-web` are incompatible. Please see my suggestions which at least make it compile and run. In addition to that we should also have a test to verify these changes.

**Comment by wking** [comment]
Pushed a8733ab -> c471c49 with @steveeJ's [suggested][1] [changes][2].

[1]: https://github.com/openshift/cincinnati/pull/196#discussion_r373130239
[2]: https://github.com/openshift/cincinnati/pull/196#discussion_r373130326

File: policy-engine/src/main.rs
Comment (wking) [reply policy-engine/src/main.rs]: Applied with a8733ab -> c471c49.

File: policy-engine/Cargo.toml
Comment (wking) [reply policy-engine/Cargo.toml]: Applied with a8733ab -> c471c49.

**Comment by wking** [comment]
>  In addition to that we should also have a test to verify these changes.

Not sure how to set this up.  Would I plug into [here][1] somehow?  Looks like at the moment all that has access to is the body of the response, not the full HTTP response with headers and a body.

[1]: https://github.com/openshift/cincinnati/blob/1207dda130fcdb0a2cc7014a508927103519adf3/policy-engine/src/graph.rs#L387

**Comment by steveej** [comment]
I think you forgot to push the changes to the `Cargo.lock` after my recent suggestions.

I played with the tests locally and would like to add a commit on top of your changes if that's fine with you. The tests will still be broken but we can continue to work on it together based on my commit.

**Comment by wking** [comment]
> I think you forgot to push the changes to the `Cargo.lock`...

openshift/release#6935 is in flight to make this easier to do, and impossible to forget (because the required `images` CI job would fail).

**Review by LalatenduMohanty — CHANGES_REQUESTED** [review]
Why do not we have any tests with this PR?

**Comment by steveej** [comment]
I've been working on getting tests in for this as well as figuring out if this is working correctly in this branch: https://github.com/steveeJ-forks/openshift-cincinnati/tree/forks/wking/cors

However, I can't get the policy-engine to respond to or return any CORS related headers. I suggest we wait for #195 to land which bumps all actix crates to their latest versions. The latest versions have better documentation and we might have more luck with that.

On a side note, I don't fully understand the issue any client would be seeing without this PR. Cincinnati doesn't block any requests for me, no matter what `Origin` header I provide with the request.

**Comment by wking** [comment]
> On a side note, I don't fully understand the issue any client would be seeing without this PR.

CORS is a client-side mechanism.  More details in the MDN page linked from my initial PR post, but basically this is Cincy telling browsers "it's ok if a page outside of api.openshift.com to make Cincy requests.  That's neither a XSS attack nor malicious data exfiltration.".

**Comment by LalatenduMohanty** [comment]
> CORS is a client-side mechanism. More details in the MDN page linked from my initial PR post, but basically this is Cincy telling browsers "it's ok if a page outside of api.openshift.com to make Cincy requests. That's neither a XSS attack nor malicious data exfiltration.".

+1 , basically with this PR we are open to CORS requests from clients which would fail otherwise. AFAIK the typical example is our web browsers which does not allow CORS requests by default.

**Comment by wking** [comment]
Setting this header will avoid [the][1]:

    CORS header ‘Access-Control-Allow-Origin’ missing

error mentioned in 066d1dc2a9fc6585, when Cincinnati is fetched from a page outside of `api.openshift.com`.

[1]: https://github.com/openshift/cincinnati/commit/4b2493165bdaf67e8a0b69319323c37c3315e6b2#diff-8a95725022f81408c83e60b8d9ec03feR41

**Comment by steveej** [comment]
And should `Access-Control-Allow-Origin` be included in the headers in every `GET` or `HEAD` request made to any endpoint?

**Comment by wking** [comment]
> And should Access-Control-Allow-Origin be included in the headers in every GET or HEAD request made to any endpoint?

s/request/response/, but sure, why not.  Seems academic until we have other endpoints to talk about?  I don't care one way or the other about having the header on metrics responses, and don't know if the global setup I have here will affect them or not, since they're on a different port.

**Comment by kunyan** [comment]
Hey @wking , could you please rebase the code and fix the test ?

**Comment by wking** [comment]
[e2e][1]:

```
failures:
    check_slo::_cincinnati_gb_graph_upstream_errors_total_0_
    check_slo::_clamp_min_histogram_quantile_0_90_sum_cincinnati_pe_v1_graph_serve_duration_seconds_bucket_by_le_1_1_
    check_slo::_max_over_time_cincinnati_gb_graph_upstream_errors_total_1h_0_
    check_slo::_max_over_time_cincinnati_pe_http_upstream_errors_total_1h_0_
    check_slo::_r_min_over_time_up_job_cincinnati_policy_engine_1h_1_
```

I dunno how that would be related, so see if it's consistent:

/retest

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cincinnati/196/pull-ci-openshift-cincinnati-master-e2e/1362956674339442688#1:build-log.txt%3A680

**Comment by wking** [comment]
SLO overshoot [again][1].

/retest

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cincinnati/196/pull-ci-openshift-cincinnati-master-e2e/1362987167441227776

**Review by vrutkovs — APPROVED** [review]
/lgtm

**Comment by wking** [comment]
/hold

[More SLO overshoot][1]:

```
failures:
    check_slo::_clamp_min_histogram_quantile_0_90_sum_cincinnati_pe_v1_graph_serve_duration_seconds_bucket_by_le_1_1_
    check_slo::_max_over_time_cincinnati_pe_http_upstream_errors_total_1h_0_
    check_slo::_r_min_over_time_up_job_cincinnati_policy_engine_1h_1_
```

I think we need to figure out what's going on with the delay before we can land this change.

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cincinnati/196/pull-ci-openshift-cincinnati-master-e2e/1363953724292075520#1:build-log.txt%3A567

**Comment by wking** [comment]
Poking at the recent job:

* `_max_over_time_cincinnati_pe_http_upstream_errors_total_1h_0_` [failed][1], expecting 0, got `0: the vector contains 0 elements`.
* `_r_min_over_time_up_job_cincinnati_policy_engine_1h_1_` [failed][2], expecting 1, got 0.
* `_clamp_min_histogram_quantile_0_90_sum_cincinnati_pe_v1_graph_serve_duration_seconds_bucket_by_le_1_1_` [failed][3], expecting 0, got `0: the vector contains 0 elements`.

But there's nothing alarming in the [policy-engine][4] [logs][5].  I dunno...

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cincinnati/196/pull-ci-openshift-cincinnati-master-e2e/1363953724292075520#1:build-log.txt%3A316
[2]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cincinnati/196/pull-ci-openshift-cincinnati-master-e2e/1363953724292075520#1:build-log.txt%3A430
[3]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cincinnati/196/pull-ci-openshift-cincinnati-master-e2e/1363953724292075520#1:build-log.txt%3A373
[4]: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cincinnati/196/pull-ci-openshift-cincinnati-master-e2e/1363953724292075520/artifacts/e2e/gather-extra/artifacts/pods/cincinnati-e2e_cincinnati-1-d7sqm_cincinnati-policy-engine.log
[5]: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cincinnati/196/pull-ci-openshift-cincinnati-master-e2e/1363953724292075520/artifacts/e2e/gather-extra/artifacts/pods/cincinnati-e2e_cincinnati-1-p9sln_cincinnati-policy-engine.log

**Comment by wking** [comment]
I dunno.  I'm going to pull the hold and let the retest bot bang at this some more while I think ;).

/hold cancel

**Comment by wking** [comment]
```context
$ curl -s 'https://search.ci.openshift.org/search?maxAge=24h&type=junit&name=pull-ci-openshift-cincinnati-master-e2e&search=failures:$&context=7' | jq -r 'to_entries[] | .key as $k | .value | to_entries[].value[] | $k + "\n> " + (.context | jo
in("\n> "))'
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cincinnati/196/pull-ci-openshift-cincinnati-master-e2e/1363794811584253952
> ...
> failures:
>     check_slo::_cincinnati_gb_graph_upstream_errors_total_0_
>     check_slo::_clamp_min_histogram_quantile_0_90_sum_cincinnati_pe_v1_graph_serve_duration_seconds_bucket_by_le_1_1_
>     check_slo::_max_over_time_cincinnati_gb_graph_upstream_errors_total_1h_0_
>     check_slo::_max_over_time_cincinnati_pe_http_upstream_errors_total_1h_0_
>     check_slo::_r_min_over_time_up_job_cincinnati_policy_engine_1h_1_
> 
> test result: FAILED. 2 passed; 5 failed; 0 ignored; 0 measured; 0 filtered out
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cincinnati/196/pull-ci-openshift-cincinnati-master-e2e/1363829442085916672
> ...
> failures:
>     check_slo::_clamp_min_histogram_quantile_0_90_sum_cincinnati_pe_v1_graph_serve_duration_seconds_bucket_by_le_1_1_
>     check_slo::_max_over_time_cincinnati_pe_http_upstream_errors_total_1h_0_
>     check_slo::_r_min_over_time_up_job_cincinnati_policy_engine_1h_1_
> 
> test result: FAILED. 4 passed; 3 failed; 0 ignored; 0 measured; 0 filtered out
> ...
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cincinnati/196/pull-ci-openshift-cincinnati-master-e2e/1363849081264279552
> ...
> failures:
>     check_slo::_clamp_min_histogram_quantile_0_90_sum_cincinnati_pe_v1_graph_serve_duration_seconds_bucket_by_le_1_1_
>     check_slo::_max_over_time_cincinnati_pe_http_upstream_errors_total_1h_0_
>     check_slo::_r_min_over_time_up_job_cincinnati_policy_engine_1h_1_
> 
> test result: FAILED. 4 passed; 3 failed; 0 ignored; 0 measured; 0 filtered out
> ...
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cincinnati/196/pull-ci-openshift-cincinnati-master-e2e/1363869073837920256
> ...
> failures:
>     check_slo::_clamp_min_histogram_quantile_0_90_sum_cincinnati_pe_v1_graph_serve_duration_seconds_bucket_by_le_1_1_
>     check_slo::_max_over_time_cincinnati_pe_http_upstream_errors_total_1h_0_
>     check_slo::_r_min_over_time_up_job_cincinnati_policy_engine_1h_1_
> 
> test result: FAILED. 4 passed; 3 failed; 0 ignored; 0 measured; 0 filtered out
> ...
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cincinnati/196/pull-ci-openshift-cincinnati-master-e2e/1363891516279361536
> ...
> failures:
>     check_slo::_clamp_min_histogram_quantile_0_90_sum_cincinnati_pe_v1_graph_serve_duration_seconds_bucket_by_le_1_1_
>     check_slo::_max_over_time_cincinnati_pe_http_upstream_errors_total_1h_0_
>     check_slo::_r_min_over_time_up_job_cincinnati_policy_engine_1h_1_
> 
> test result: FAILED. 4 passed; 3 failed; 0 ignored; 0 measured; 0 filtered out
> ...
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cincinnati/196/pull-ci-openshift-cincinnati-master-e2e/1363953724292075520
> ...
> failures:
>     check_slo::_clamp_min_histogram_quantile_0_90_sum_cincinnati_pe_v1_graph_serve_duration_seconds_bucket_by_le_1_1_
>     check_slo::_max_over_time_cincinnati_pe_http_upstream_errors_total_1h_0_
>     check_slo::_r_min_over_time_up_job_cincinnati_policy_engine_1h_1_
> 
> test result: FAILED. 4 passed; 3 failed; 0 ignored; 0 measured; 0 filtered out
> ...
```

I dunno where the graph-builder failures went, but the policy-engine failures seem pretty stable :/

**Comment by vrutkovs** [comment]
/retest

probe timeout was increased

**Comment by vrutkovs** [comment]
Lets

/hold

it for > 4hrs to make it recreate a namespace

**Comment by vrutkovs** [comment]
Needs rebase due to Cargo.toml changes

**Comment by wking** [comment]
Rebased around #371 and #360, but [still][1] overshooting the SLOs.  I think I'm probably holding the CORS library wrong...

```
failures:
    check_slo::_clamp_min_histogram_quantile_0_90_sum_cincinnati_pe_v1_graph_serve_duration_seconds_bucket_by_le_1_1_
    check_slo::_max_over_time_cincinnati_pe_http_upstream_errors_total_1h_0_
    check_slo::_r_min_over_time_up_job_cincinnati_policy_engine_1h_1_
```

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cincinnati/196/pull-ci-openshift-cincinnati-master-e2e/1365161646162448384#1:build-log.txt%3A490

**Comment by vrutkovs** [comment]
Still seeing readiness probe failures

/retest

**Comment by vrutkovs** [comment]
Timeline:
> 10:29:03 Successfully assigned cincinnati-e2e/cincinnati-1-ddw22 to ip-10-0-166-203.ec2.internal" 

14 seconds for process in the container to start:
> 10:29:17 DEBUG graph_builder] application settings:
> 10:29:48 Readiness probe failed: HTTP probe failed with statuscode: 503"
> 10:29:51 DEBUG graph_builder::graph] graph update completed

So readiness probe doesn't account for the time for binary in the container to start. Its weird that it takes that to start though

File: policy-engine/src/main.rs
Comment (vrutkovs) [inline policy-engine/src/main.rs]: I think we want [allow_any_origin](https://docs.rs/actix-cors/0.5.4/actix_cors/struct.Cors.html#method.allow_any_origin) here. See also https://docs.rs/actix-cors/0.5.4/src/actix_cors/builder.rs.html#140

Comment (wking) [reply policy-engine/src/main.rs]: Thanks; fixed with 8e4ca504 -> 95c923cd

**Review by vrutkovs — APPROVED** [review]
/lgtm

**Pull request merged**
